### ISSUE TITLE
Feature Request: loading ffi shared libs from Uint8Array directly

### ISSUE THREAD

**Issue by andykais** [denoland/deno#15700]
The idea here is pretty simple, in addition to `Deno.dlopen` accepting a `string` or `URL`, lets also let it accept a `Uint8Array`. The reason for this is so we can start embedding shared libraries in javascript files, so that they can be downloaded at import time. This idea is already widely used with wasm binaries, so in my opinion it isnt a large stretch for ffi libraries either

**Comment by eliassjogreen** [comment]
Loading of dynamic libraries from memory is not natively supported, at least not on unix and is quite hard to implement as there are no pre-made solutions for rust that I know of. The solution would be to transport the dll as a javascript file, much like is done with wasm and then to write it to a file manually and load. Relevant [StackOverflow](https://stackoverflow.com/a/5053703) question about the same thing.

TLDR; not really possible

**Comment by piscisaureus** [comment]
On linux it could be done with memfd_create().
How to do it on OS X and Windows is an open question though.

**Comment by andykais** [comment]
what @eliassjogreen outlined is what I ended up solutioning. In my case, this is an experiment for sqlite, so read & write permissions are already used by default, so its not a big deal. In other ffi cases though, it would probably be more awkward to add write permissions just to write an embedded file.

I did not realize loading dynamic libs by file was done all the way down the stack, that does make this a pretty untenable idea. I can close this issue if its not likely to be solved

**Comment by piscisaureus** [comment]
OS X: https://stackoverflow.com/a/11822029
Windows: https://github.com/fancycode/MemoryModule

**Comment by littledivy** [comment]
On macOS, NSCreateObjectFileImageFromMemory is a long-depreacted API and a has been a source of code loading attacks (https://hackd.net/posts/macos-reflective-code-loading-analysis/)

Apple changed it so It simply writes to disk and uses file loading. https://github.com/apple-oss-distributions/dyld/blob/3f24a36068a96722cf3acbd5087983ce658e9d70/dyld3/APIs_macOS.cpp#L154

**Comment by bartlomieju** [comment]
Closing as not planned.

The underlying OS primitives are inherently path-based: \`Deno.dlopen\` maps to
\`dlopen(3)\` on Unix and \`LoadLibrary\` on Windows (via \`dlopen2\`'s
\`Library::open(path)\` in \`ext/ffi/dlfcn.rs\`), and there is no portable way to
load a shared library directly from memory -- the Linux \`memfd\`/\`/proc/self/fd\`
trick, Windows manual PE loaders, and the (removed) macOS in-memory image APIs
are all non-portable and/or unsupported.

So a \`Uint8Array\` overload couldn't avoid touching disk -- internally it would
have to write the bytes to a temp file and \`dlopen\` that path. That's something
you can already do yourself in a few lines, with control over location and
cleanup:

\`\`\`ts
const path = await Deno.makeTempFile({ suffix: ".so" }); // or .dll/.dylib
await Deno.writeFile(path, libBytes);
const lib = Deno.dlopen(path, symbols);
// ... use lib; remove the temp file when done
\`\`\`

Beyond the portability problem, a buffer-based load would also bypass the
path-based FFI permission model (\`--allow-ffi=<path>\`), where granting access
to a specific library on disk is the unit of trust. Given there's no real
abstraction win over the snippet above and it has security/portability
downsides, this isn't something we plan to add.
